baton/modules/twilio/twilio.go

```diff
@@ -0,0 +1,179 @@
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	var out map[string]interface{}
+	if err := json.NewDecoder(res.Body).Decode(&out); err != nil {
+		return err
+	}
+	delete(out, "account_sid")
+	resp <- out
+	return nil
+}
+
+type callXml struct {
+	RestException twilioError
+}
+type twilioError struct {
+	Code     int
+	Message  string
+	MoreInfo string
+	Status   int
+}
+
+func (t Twilio) makeCall(body map[string]interface{}, resp chan<- interface{}) error {
+	to := body["to"].(string)
+	from := body["from"].(string)
+	twiml := body["twiml"].(string)
+	form := url.Values{"To": {to}, "From": {from}, "Url": {"http://e921edd9.ngrok.io/webhooks/Twilio/call"}} //temporary
+	callCallbacks = append(callCallbacks, callback{to, resp, twiml})
+	res, err := t.postForm(fmt.Sprintf("https://api.twilio.com/2010-04-01/Accounts/%s/Calls", t.UserId), form)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+	var out callXml
+	if err := xml.NewDecoder(res.Body).Decode(&out); err != nil {
+		return err
+	}
+	if out.RestException.Code != 0 {
+		return errors.New(out.RestException.Message)
+	}
+	return nil
+}
+func (t Twilio) recieveSMS(body map[string]interface{}, resp chan<- interface{}) error {
+	from := body["from"].(string)
+	smsCallbacks = append(smsCallbacks, callback{from, resp, ""})
+	return nil
+}
+
+func (t Twilio) recieveCall(body map[string]interface{}, resp chan<- interface{}) error {
+	from := body["from"].(string)
+	twiml := body["twiml"].(string)
+	callCallbacks = append(callCallbacks, callback{from, resp, twiml})
+	return nil
+}
+func (t Twilio) postForm(url string, form url.Values) (*http.Response, error) {
+	req, err := http.NewRequest("POST", url, strings.NewReader(form.Encode()))
+	if err != nil {
+		return nil, err
+	}
+	req.SetBasicAuth(t.UserId, t.ApiKey)
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func (t Twilio) Handler() *mux.Router {
+	m := mux.NewRouter()
+
+	m.Path("/sms").HandlerFunc(sms)
+	m.Path("/call").HandlerFunc(call)
+	return m
+}
+func sms(w http.ResponseWriter, r *http.Request) {
+	err := r.ParseForm()
+	if err != nil {
+		fmt.Println(err)
+	}
+	out := make(map[string]string)
+	for name, val := range r.PostForm {
+		out[name] = val[0]
+	}
+	delete(out, "AccountSid")
+	for _, callback := range smsCallbacks {
+		if callback.number == out["From"] || callback.number == "*" {
+			callback.resp <- out
+		}
+	}
+
+	fmt.Println(out)
+}
+func call(w http.ResponseWriter, r *http.Request) {
+	err := r.ParseForm()
+	if err != nil {
+		return
+	}
+	out := make(map[string]string)
+	for name, val := range r.PostForm {
+		out[name] = val[0]
+	}
+	delete(out, "AccountSid")
+	for i, callback := range callCallbacks {
+		if "inbound" == out["Direction"] {
+			if callback.number == out["Caller"] {
+				fmt.Fprintf(w, "<?xml version=\"1.0\" encoding=\"UTF-8\"?><Response>%s</Response>", callback.data)
```

Comment:
Currently we are generating xml on the client-side; in order to encode using `encoding/xml`, we would send something like below from the client, 
```
twiml:[{type:"pause",value:5},
{type:"say",value:"hello world"},
{type:"play",value:"mp3 url"}
]
```
And then encode that into xml on the server.